### DISCUSSION TITLE
Create Error Surface For Entire DEM Extent/ FIS error model

### DISCUSSION THREAD

**Discussion by NatBarbosa** [Riverscapes/gcd#416] [I need help please] [answered]
## The Problem

Hi all, 

I am not able to create an error surface for an entire DEM using a FIS model with two parameters, i.e., Slope degree and Point Density. I have tried different pre-build models such CHaMP_TS_ZError_SLPdeg_PD_2012, GPS_ZError_PD_SLPdeg, TS_ZError_PD_SLPdeg. 

In the future, I would like to implement the method using my own model as my aim is to estimate the spatial Zerror from photogrammetric DEMs.

## Reproduction steps

1. Load the DEM survey
2. Calculate associated Surfaces
  2.1. Slope Degrees
  2.2. Point Density from a point shapefile
3. Load associated Surfaces
  3.1. Point density computed with Cloud Compare at the same spatial extent and resolution of the DEM Survey
4. Error Surfaces 
  4.1. Calculate new error surface 
         4.1.1. Uniform error value - successful 
        4.1.2. FIS error model -- PROBLEM OCCURRED 

I tried the same workflow in both the standalone version and ArcMap plugin. 
I tried the same workflow with my own dataset and the GeoTERM_Feshie and N_FIS_Intro datasets provided on the webpage. 

## Exception message

**GCD 7 Standalone**

'Error editing single region error surface' 

Membership function vertices (1 25 75 1) must be in ascending order.
 --- Stacktrace --- 
   at GCDConsoleLib.FIS.MemberFunction..ctor(Double x1, Double x2, Double x3, Double x4, Double dyMax)
   at GCDConsoleLib.FIS.FisFile.parseInputOutput(List`1 sLines, Boolean isInput)
   at GCDConsoleLib.FIS.FisFile..ctor(FileInfo inputFn)
   at GCDConsoleLib.Internal.Operators.FISRasterOp..ctor(Dictionary`2 rInputs, FileInfo fisFile)
   at GCDConsoleLib.Internal.Operators.CreateErrorRaster..ctor(Raster rawDEM, ErrorRasterProperties prop, Raster rOutputRaster)
   at GCDConsoleLib.RasterOperators.CreateErrorRaster(Raster rawDEM, ErrorRasterProperties props, FileInfo sOutputRaster, EventHandler`1 progressHandler)
   at GCDCore.UserInterface.SurveyLibrary.ErrorSurfaces.frmSingleMethodError.cmdOK_Click(Object sender, EventArgs e)
 --- Exception Data --- 
GCD = 7.5.0.0
Windows: Microsoft Windows NT 6.2.9200.0
Date: 25/08/2021 09:31:40

```

## Anything else?

You can attach links to datasets.

**Comment by joewheaton** [comment]
Sorry to hear about this @NatBarbosa. If my explanation is off, please upload your *.fis file or copy the contents in as code to this post. 

The error message `'Error editing single region error surface Membership function vertices (1 25 75 1) must be in ascending order.` is actually quite helpful. I am not sure which Membership function this is, but the values sound like degrees slope and perhaps your high one. If it really came from  [`GPS_ZError_PD_SLPdeg.fis`](https://github.com/Riverscapes/fis-dem-error/blob/master/BySurveyType/GPS/GPS_ZError_PD_SLPdeg.fis), then presumably that was an edit of either the slope or point density? 

Assuming what you edited was something like:
`MF2='Medium':'trapmf',[1 25 75 1]`
Notice in original the values are ascending (i.e. 100 is > 75 whereas in yours 1 IS NOT > 75)
https://github.com/Riverscapes/fis-dem-error/blob/e8c54fd0fcb0f6495c04aeee3f41883f9f726ea2/BySurveyType/GPS/GPS_ZError_PD_SLPdeg.fis#L27

It needs to be ascending. I explain what is going wrong in this rambling [video](https://youtu.be/agMvv0YL8hw):

[![](http://img.youtube.com/vi/agMvv0YL8hw/0.jpg)](http://www.youtube.com/watch?v=agMvv0YL8hw"")

**Reply by NatBarbosa** [reply]
Dear @joewheaton, 

Thank you so much for such a complete and fast answer. Unfortunately, I keep experiencing the same error after following closely your video using the same dataset. After exploring in deep the definition of the FSI I recognized that the problem might be related to the decimal symbol!!

I am currently based in Germany and the standard decimal symbol here is a ",". I have changed this in the operating system configuration to correspond to a '.'. I am able to get the expected results after this change! 

I got this idea from the definition of:

[Input1]
Name='SlopeDeg'
Range=[0 90]
NumMFs=3
MF1='Low':'trapmf',[0 0 2.8 5.7]
MF2='Medium':'trapmf',[2.8 5.7 8.5 11.3]
MF3='High':'trapmf',[8.5 11.3 90 90]  

The values in the MF3 correspond to the error message if the '.' is not interpreted as a decimal indicator ---> (85 113 90 90)

I hope this information be useful for other users and developers!

Kind regards, 

Natalie.

**Reply by joewheaton** [reply]
Nice find. Yes, the commas came up previously (not for FIS, but for other user interface stuff) and we had to figure that out. We just adopted Matlab's Fuzzy Logic Toolbox format and convention here.

**Comment by joewheaton** [comment, marked as answer]
Good catch @NatBarbosa. Although customary in many European countries to substitute a `,` comma in place of a `.` for a decimal, it will break the FIS (we inherit this from Matlab Fuzzy Logic Toolbox).
